feat: Add global event bus

flipswitchingmonkey opened the pull request.

flipswitchingmonkey committed:
fix branch

flipswitchingmonkey committed:
fix deserialize, add filewriter

flipswitchingmonkey committed:
add catchAll eventGroup/Name

flipswitchingmonkey committed:
adding simple Redis sender and receiver to eventbus

flipswitchingmonkey committed:
remove native node threads

flipswitchingmonkey committed:
improve eventbus

flipswitchingmonkey committed:
refactor and simplify

flipswitchingmonkey committed:
more refactoring and syslog client

flipswitchingmonkey committed:
more refactor, improved endpoints and eventbus

flipswitchingmonkey committed:
remove local broker and receivers from mvp

flipswitchingmonkey committed:
destination de/serialization

flipswitchingmonkey committed:
create MessageEventBusDestinationEntity

flipswitchingmonkey committed:
db migrations, load destinations at startup

flipswitchingmonkey committed:
add delete destination endpoint

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	package-lock.json
#	packages/cli/package.json
#	packages/cli/src/Interfaces.ts
#	packages/cli/src/Server.ts

flipswitchingmonkey committed:
pnpm merge and circular import fix

flipswitchingmonkey committed:
delete destination fix

flipswitchingmonkey committed:
trigger log file shuffle after size reached

flipswitchingmonkey committed:
add environment variables for eventbus

flipswitchingmonkey committed:
reworking event messages

flipswitchingmonkey committed:
serialize to thread fix

flipswitchingmonkey committed:
some refactor and lint fixing

flipswitchingmonkey committed:
add emit to eventbus

flipswitchingmonkey committed:
cleanup and fix sending unsent

flipswitchingmonkey committed:
quicksave frontend trial

flipswitchingmonkey committed:
initial EventTree vue component

flipswitchingmonkey committed:
basic log streaming settings in vue

flipswitchingmonkey committed:
http request code merge

flipswitchingmonkey committed:
create destination settings modals

flipswitchingmonkey committed:
fix eventmessage options types

flipswitchingmonkey committed:
credentials are loaded

flipswitchingmonkey committed:
fix and clean up frontend code

flipswitchingmonkey committed:
move request code to axios

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
update lock file

flipswitchingmonkey committed:
merge fix

flipswitchingmonkey committed:
fix redis build

flipswitchingmonkey committed:
move destination interfaces into workflow pkg

flipswitchingmonkey committed:
revive sentry as destination

flipswitchingmonkey committed:
migration fixes and frontend cleanup

flipswitchingmonkey committed:
N8N-5777 / N8N-5789 N8N-5788

flipswitchingmonkey committed:
N8N-5784

flipswitchingmonkey committed:
N8N-5782 removed event levels

flipswitchingmonkey committed:
N8N-5790 sentry destination cleanup

flipswitchingmonkey committed:
N8N-5786 and refactoring

flipswitchingmonkey committed:
N8N-5809 and refactor/cleanup

flipswitchingmonkey committed:
UI fixes and anonymize renaming

flipswitchingmonkey committed:
N8N-5837

flipswitchingmonkey committed:
N8N-5834

flipswitchingmonkey committed:
fix no-items UI issues

flipswitchingmonkey committed:
remove card / settings label in modal

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	packages/cli/src/commands/start.ts

flipswitchingmonkey committed:
N8N-5842 fix

flipswitchingmonkey committed:
disable webhook auth for now and update ui

flipswitchingmonkey committed:
change sidebar to tabs

flipswitchingmonkey committed:
remove payload option

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	packages/cli/package.json
#	packages/cli/src/Interfaces.ts
#	packages/cli/src/WorkflowExecuteAdditionalData.ts
#	packages/cli/src/databases/entities/index.ts
#	packages/cli/src/databases/migrations/mysqldb/index.ts
#	packages/cli/src/databases/migrations/postgresdb/index.ts
#	packages/cli/src/databases/migrations/sqlite/index.ts

flipswitchingmonkey committed:
extend audit events with more user data

flipswitchingmonkey committed:
N8N-5853 and UI revert to sidebar

flipswitchingmonkey committed:
remove redis destination

flipswitchingmonkey committed:
N8N-5864 / N8N-5868 / N8N-5867 / N8N-5865

flipswitchingmonkey committed:
ui and licensing fixes

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	packages/cli/src/Server.ts

flipswitchingmonkey committed:
add node events and info bubbles to frontend

flipswitchingmonkey committed:
ui wording changes

flipswitchingmonkey committed:
frontend tests

flipswitchingmonkey committed:
N8N-5896 and ee rename

flipswitchingmonkey committed:
improves backend tests

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	packages/cli/src/config/schema.ts
#	packages/editor-ui/src/components/MainSidebar.vue
#	packages/editor-ui/src/components/Modals.vue
#	packages/editor-ui/src/components/SettingsSidebar.vue
#	packages/editor-ui/src/constants.ts
#	packages/editor-ui/src/mixins/genericHelpers.ts

krynble commented on packages/cli/src/config/schema.ts:
```diff
@@ -1012,6 +1016,53 @@ export const schema = {
+			default: true,
+			env: 'N8N_EVENTBUS_ENABLED',
+		},
+		checkUnsentInterval: {
```
What are the pros and cons of this behavior? Are we documenting this somewhere? (A docs ticket or something to assist in deciding when to turn this on)

flipswitchingmonkey replied:
agreed - there isn't much use to turn it off, but possible side effects if it is _not_ running. so I'll remove this option (a leftover from earlier iterations...)

krynble commented on packages/cli/src/config/schema.ts:
```diff
@@ -1012,6 +1016,53 @@ export const schema = {
+			env: 'N8N_EVENTBUS_CHECKUNSENTINTERVAL',
+		},
+		destinations: {
+			loadAtStart: {
+				doc: 'Whether destinations are loaded from Db and initialized at startup.',
+				format: Boolean,
+				default: true,
+				env: 'N8N_EVENTBUS_DESTINATIONS_STARTUP',
```
Why would I choose not to load from bd? I guess once registered I expect n8n to always load this automatically.

flipswitchingmonkey replied:
same as above. was an early option and will now be removed.

krynble commented on packages/cli/src/config/schema.ts:
```diff
@@ -1012,6 +1016,53 @@ export const schema = {
+			maxFileSizeInKB: {
+				doc: 'Maximum size of an event log file before a new one is started.',
+				format: Number,
+				default: 102400,
```
```suggestion
				default: 102400, // 100kb
```

flipswitchingmonkey replied:
it's in KB, so this is 100MB. but still makes sense to add a comment to avoid this confusion :)

krynble commented on packages/cli/src/config/schema.ts:
```diff
@@ -1012,6 +1016,53 @@ export const schema = {
+				env: 'N8N_EVENTBUS_DESTINATIONS_STARTUP',
+			},
+		},
+		logWriter: {
```
Do you think it makes sense to merge those with the other logging settings?

flipswitchingmonkey replied:
it's very much hard-linked with the eventbus, since it is what's launching the process with the log writer (unlike the loggerproxy, which I think runs through Watson). Since both are configuring totally different objects and logs, I'd keep them apart.

flipswitchingmonkey committed:
merge fix

krynble commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageSerialized.ts:
```diff
@@ -0,0 +1,13 @@
+// import { AbstractEventPayload } from './AbstractEventPayload';
+// import { EventMessageOptions } from './EventMessageOptions';
+
+// // export interface EventMessageSerialized extends Required<Omit<EventMessageOptions, 'payload'>> {
```
If not necessary, can we remove the file?

krynble commented on packages/cli/src/eventbus/MessageEventBus/MessageEventBus.ts:
```diff
@@ -0,0 +1,259 @@
+		[key: string]: MessageEventBusDestination;
+	} = {};
+
+	#pushInteralTimer: NodeJS.Timer;
```
```suggestion
	#pushIntervalTimer: NodeJS.Timer;
```

flipswitchingmonkey committed:
fix backend test

krynble commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestinationSentry.ee.ts:
```diff
@@ -0,0 +1,131 @@
+					environment,
+					release: versions.cli,
+				});
+				console.debug(`MessageEventBusDestinationSentry with id ${this.getId()} initialized`);
```
Can we change to `LoggerProxy` instead of the `console` for logging? It standardizes the default log output

krynble commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestinationSentry.ee.ts:
```diff
@@ -0,0 +1,131 @@
+
+		GenericHelpers.getVersions()
+			.then((versions) => {
+				Sentry.init({
```
Is it possible to set up two Sentry accounts for uploading data? What would happen then? Is it possible to keep two separate Sentry account contexts?

flipswitchingmonkey replied:
Fixed!

krynble commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestinationSyslog.ee.ts:
```diff
@@ -0,0 +1,146 @@
+		this.eol = options.eol ?? '\n';
+		this.expectedStatusCode = options.expectedStatusCode ?? 200;
+
+		this.client = syslog.createClient(this.host, {
```
Have we considered using the already installed `winston` library?

flipswitchingmonkey replied:
there's a winston syslog extension but it did not work reliably for some reason (it was rather old as well...)

krynble commented on packages/cli/src/eventbus/eventBusRoutes.ts:
```diff
@@ -0,0 +1,236 @@
+);
+
+eventBusRouter.post(
+	`/event/addmany/:count`,
```
```suggestion
	`/event/test/addmany/:count`,
```
Considering this is a test endpoint, I think it's worth adding `test` as a URI segment to state the purpose

flipswitchingmonkey replied:
i'll remove it entirely... it wasn't meant to even be there still.

krynble commented on packages/cli/src/eventbus/eventBusRoutes.ts:
```diff
@@ -0,0 +1,236 @@
+	}),
+);
+
+eventBusRouter.post(
```
Shouldn't those endpoints be protected to be accessible by instance owner only?

flipswitchingmonkey replied:
will change both back and frontend to be only available to owners (for change actions, reading available for all)

flipswitchingmonkey committed:
make linter happy

flipswitchingmonkey commented on packages/cli/src/InternalHooks.ts:
```diff
@@ -15,8 +15,27 @@ import {
+
+function userToPayload(user: User): {
+	userId: string;
+	_email: string;
+	_firstName: string;
+	_lastName: string;
```
should be fine, it's just left null/undefined (and should not show up in the payload)

flipswitchingmonkey commented on packages/cli/src/InternalHooks.ts:
```diff
@@ -15,8 +15,27 @@ import {
 } from '@/Interfaces';
 import { Telemetry } from '@/telemetry';
+import { eventBus } from './eventbus';
+import { User } from './databases/entities/User';
```
as discussed, this is a larger refactor, but makes sense in the future (merging events, telemetry and statistics)

flipswitchingmonkey commented on packages/cli/src/WorkflowExecuteAdditionalData.ts:
```diff
@@ -639,6 +640,17 @@ function hookFunctionsSave(parentProcessMode?: string): IWorkflowExecuteHooks {
 					});
+					Promise.all([
+						eventBus.sendWorkflowEvent({
+							eventName: 'n8n.workflow.error',
```
👍

flipswitchingmonkey commented on packages/cli/src/WorkflowRunner.ts:
```diff
@@ -51,6 +51,7 @@ import { InternalHooksManager } from '@/InternalHooksManager';
 import { generateFailedExecutionFromError } from '@/WorkflowHelpers';
 import { initErrorHandling } from '@/ErrorReporting';
 import { PermissionChecker } from '@/UserManagement/PermissionChecker';
+import { eventBus } from './eventbus';
```
👍

flipswitchingmonkey committed:
remove unnecessary cfg / limit  actions to owners

flipswitchingmonkey committed:
fix multiple sentry DSN and anon bug

flipswitchingmonkey committed:
eslint fix

flipswitchingmonkey committed:
more tests and fixes

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

# Conflicts:
#	packages/editor-ui/src/mixins/workflowHelpers.ts

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
merge fix

netroy committed:
Merge remote-tracking branch 'origin/master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
fix workflow audit events

flipswitchingmonkey committed:
remove 'n8n.workflow.execution.error' event

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
merge fix

flipswitchingmonkey committed:
lint fix

krynble commented on packages/cli/src/Server.ts:
```diff
@@ -388,6 +394,7 @@ class App {
 		// refresh enterprise status
 		Object.assign(this.frontendSettings.enterprise, {
 			sharing: isSharingEnabled(),
+			logStreaming: isLogStreamingEnabled(),
```
Could you please add this setting to the `IN8nUISettings` interface in `packages/cli/src/Interfaces.ts`?

krynble commented on packages/cli/src/Server.ts:
```diff
@@ -1539,6 +1546,20 @@ class App {
+		// add Event Bus REST endpoints
+		this.app.use(`/${this.restEndpoint}/eventbus`, eventBusRouter);
+
+		if (process.env.NODE_ENV !== 'production') {
+			config.set('enterprise.features.logStreaming', true);
+		}
+
```
Should we remove this before merging?

flipswitchingmonkey replied:
we can. it just makes testing a bit harder.

krynble commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestinationSentry.ee.ts:
```diff
@@ -0,0 +1,142 @@
+		let sendResult = false;
+		if (!this.sentryInitSuccessful) return sendResult;
+		if (msg.eventName !== eventMessageGenericDestinationTestEvent) {
+			if (!isLogStreamingEnabled()) return sendResult;
```
This if is duplicated for every Destination - can we perhaps add this check to the caller?

flipswitchingmonkey replied:
This check is in there to prevent someone from sending a message to the destination directly, circumventing a disabled logStreaming

flipswitchingmonkey committed:
lint fix

flipswitchingmonkey committed:
review fixes

flipswitchingmonkey committed:
fix merge

flipswitchingmonkey committed:
prettier fixes

flipswitchingmonkey committed:
Merge branch 'master' into n8n-5359-mvp-event-bus

flipswitchingmonkey committed:
merge

netroy commented on packages/cli/src/eventbus/MessageEventBusWriter/MessageEventBusLogWriter.ts:
```diff
@@ -0,0 +1,221 @@
+import remove from 'lodash.remove';
+import config from '@/config';
+import { getEventMessageObjectByType } from '../EventMessageClasses/Helpers';
+import { EventMessageReturnMode } from '../MessageEventBus/MessageEventBus';
+import { EventMessageTypes } from '../EventMessageClasses';
```
```suggestion
import type { EventMessageReturnMode } from '../MessageEventBus/MessageEventBus';
import type { EventMessageTypes } from '../EventMessageClasses';
```

netroy commented on packages/cli/src/eventbus/MessageEventBus/MessageEventBus.ts:
```diff
@@ -0,0 +1,254 @@
+import { MessageEventBusLogWriter } from '../MessageEventBusWriter/MessageEventBusLogWriter';
+import EventEmitter from 'node:events';
+import config from '@/config';
+import { Db } from '@/index';
```
To avoid the circular dependency

```suggestion
import * as Db from '@/Db';
```

netroy commented on packages/cli/src/eventbus/MessageEventBusDestination/Helpers.ee.ts:
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable import/no-cycle */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+import { MessageEventBusDestinationTypeNames } from 'n8n-workflow';
+import { EventDestinations } from '@/databases/entities/MessageEventBusDestinationEntity';
+import { MessageEventBusDestination } from './MessageEventBusDestination.ee';
```
```suggestion
import type { EventDestinations } from '@/databases/entities/MessageEventBusDestinationEntity';
import type { MessageEventBusDestination } from './MessageEventBusDestination.ee';
```

netroy commented on packages/cli/src/databases/entities/MessageEventBusDestinationEntity.ts:
```diff
@@ -0,0 +1,12 @@
+import { Column, Entity, PrimaryColumn } from 'typeorm';
+import { AbstractEntity, jsonColumnType } from './AbstractEntity';
+
+@Entity({ name: 'eventdestinations_entity' })
```
if we rename the table to `event_destinations`, we can change this to `@Entity()`

flipswitchingmonkey replied:
_entity is added by the ORM iirc

netroy replied:
since many of our tables don't have that suffix, I believe this yet another case of copy-pasting. TypeORM doesn't have any implicit suffix.

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessage.ts:
```diff
@@ -0,0 +1,145 @@
+	abstract setPayload(payload: AbstractEventPayload): this;
+
+	anonymize(): AbstractEventPayload {
+		const anonymizedPayload = modifyUnderscoredKeys(this.payload, () => '*');
```
since the same `modifier` is used everywhere, you could use this as the default value on `modifyUnderscoredKeys`

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessage.ts:
```diff
@@ -0,0 +1,145 @@
+		this.setOptionsOrDefault(options);
+	}
+
+	// abstract serialize(): EventMessageSerialized;
```
i suppose this could be deleted.

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessageOptions.ts:
```diff
@@ -0,0 +1,12 @@
+import { DateTime } from 'luxon';
+import { AbstractEventPayload } from './AbstractEventPayload';
```
```suggestion
import type { DateTime } from 'luxon';
import type { AbstractEventPayload } from './AbstractEventPayload';
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventPayload.ts:
```diff
@@ -0,0 +1,5 @@
+import { IWorkflowBase, JsonValue } from 'n8n-workflow';
```
```suggestion
import type { IWorkflowBase, JsonValue } from 'n8n-workflow';
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageAudit.ts:
```diff
@@ -0,0 +1,74 @@
+import { AbstractEventMessage, isEventMessageOptionsWithType } from './AbstractEventMessage';
+import { EventMessageTypeNames, JsonObject, JsonValue } from 'n8n-workflow';
+import { AbstractEventPayload } from './AbstractEventPayload';
+import { AbstractEventMessageOptions } from './AbstractEventMessageOptions';
```
```suggestion
import type { AbstractEventPayload } from './AbstractEventPayload';
import type { AbstractEventMessageOptions } from './AbstractEventMessageOptions';
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageWorkflow.ts:
```diff
@@ -0,0 +1,61 @@
+
+	execution_id?: string;
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	workflow_id?: number | string | any;
```
```suggestion
	workflow_id?: number | string;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageWorkflow.ts:
```diff
@@ -0,0 +1,61 @@
+}
+
+export class EventMessageWorkflow extends AbstractEventMessage {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageNode.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+export class EventMessageNode extends AbstractEventMessage {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageGeneric.ts:
```diff
@@ -0,0 +1,43 @@
+}
+
+export class EventMessageGeneric extends AbstractEventMessage {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageConfirm.ts:
```diff
@@ -0,0 +1,41 @@
+}
+
+export class EventMessageConfirm {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy commented on packages/cli/src/eventbus/EventMessageClasses/Helpers.ts:
```diff
@@ -0,0 +1,94 @@
+	message: AbstractEventMessageOptions,
+): EventMessageTypes | null => {
+	switch (message.__type as EventMessageTypeNames) {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy commented on packages/cli/src/eventbus/EventMessageClasses/Helpers.ts:
```diff
@@ -0,0 +1,94 @@
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+		case EventMessageTypeNames.generic:
+			return new EventMessageGeneric(message as EventMessageGenericOptions);
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```
this line can be deleted

netroy reviewed (changes requested):
overall it looks good. Can you please make these minor changes, while I work on reviewing/testing this further?

flipswitchingmonkey commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessage.ts:
```diff
@@ -0,0 +1,145 @@
+	candidate: unknown,
+): candidate is AbstractEventMessageOptions => {
+	const o = candidate as AbstractEventMessageOptions;
+	if (!o) return false;
```
Not sure I understand the issue? This is just a type guard to cast the ts type to AbstractEventMessageOptions

netroy replied:
are there scenarios in the code that would make `candidate` `null` or `undefined` ?

flipswitchingmonkey replied:
not now but maybe in the future? the point is to not blindly cast the object to in this case AbstractEventMessageOptions

flipswitchingmonkey commented on packages/cli/src/eventbus/MessageEventBusWriter/MessageEventBusLogWriter.ts:
```diff
@@ -0,0 +1,221 @@
+							}
+						}
+					} catch (error) {
+						console.log(error, line);
```
Logger is intentionally not imported here because it would increase the thread's memory usage (I'm keeping the imports to a minimum)

netroy replied:
in that case can we create a small wrapper around this that respects `N8N_LOG_LEVEL` an `N8N_LOG_OUTPUT` and maybe the current Logger can extend that.

flipswitchingmonkey replied:
never mind, wrong file - this one runs in main thread so loggerproxy is ok

flipswitchingmonkey commented on packages/cli/src/InternalHooks.ts:
```diff
@@ -113,28 +151,96 @@ export class InternalHooksClass implements IInternalHooksClass {
+				},
+				{ withPostHog: true },
+			),
+		]).catch((error) => console.log(error));
```
we do not allow floating promises, thus either catch. or await.

netroy replied:
how about `return Promise.all([..])`

flipswitchingmonkey replied:
That will then result in an array of voids being returned which again will give errors

flipswitchingmonkey committed:
review changes

flipswitchingmonkey committed:
use loggerproxy

netroy committed:
Merge remote-tracking branch 'origin/master' into n8n-5359-mvp-event-bus

netroy commented:
There are multiple unexpected errors being logged in tests because of the `.catch(error => console.log(error))` lines.
![image](https://user-images.githubusercontent.com/196144/210271964-e1925c91-60ad-463f-9edb-1133711fb24a.png)

We should 
1. disable event logs in tests, except for where is makes sense
2. remove the `.catch` lines, as we can already see that they are not only catching actual promise errors but also errors from bad mocks, etc.

netroy commented:
I pushed my suggestions here https://github.com/n8n-io/n8n/pull/5073

flipswitchingmonkey committed:
remove catch from internal hook promises

flipswitchingmonkey committed:
fix tests

flipswitchingmonkey committed:
lint fix

netroy commented on packages/cli/src/InternalHooks.ts:
```diff
@@ -82,29 +104,46 @@ export class InternalHooksClass implements IInternalHooksClass {
+				public_api: publicApi,
+			}),
+		]);
+		return;
```
this `return` is implicit and not needed

netroy commented on packages/cli/src/InternalHooks.ts:
```diff
@@ -266,11 +372,40 @@ export class InternalHooksClass implements IInternalHooksClass {
 			BinaryDataManager.getInstance().persistBinaryDataForExecutionId(executionId),
 			this.telemetry.trackWorkflowExecution(properties),
-		]).then(() => {});
+		]).catch((error) => console.log(error));
```
we currently block on `persistBinaryDataForExecutionId` and if that throws, the error is propagated up. with these changes, the error now simply gets suppressed.

flipswitchingmonkey committed:
include review PR changes

netroy commented on packages/cli/src/eventbus/MessageEventBus/MessageEventBus.ts:
```diff
@@ -0,0 +1,253 @@
+		[key: string]: MessageEventBusDestination;
+	} = {};
+
+	#pushIntervalTimer: NodeJS.Timer;
```
I'm torn between staying consistent with the rest of the code using `private`, or using `#` to ensure private at runtime as well. 
The second one feels like the new semantic thing to do, but in my experience the debugging experience with private fields has been buggy and slow (likely due to the introspection calls having to go through the node's inspector api).
perhaps we could use `private pushIntervalTimer: NodeJS.Timer;` for now, and create a tech-debt ticket to switch to runtime-private fields?
WDYT?

flipswitchingmonkey replied:
moving back to private for now

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessage.ts:
```diff
@@ -0,0 +1,143 @@
+};
+
+export abstract class AbstractEventMessage {
+	abstract readonly __type: string;
```
```suggestion
	abstract readonly __type: EventMessageTypeNames;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/AbstractEventMessageOptions.ts:
```diff
@@ -0,0 +1,12 @@
+import type { AbstractEventPayload } from './AbstractEventPayload';
+
+export interface AbstractEventMessageOptions {
+	__type?: string;
```
```suggestion
	__type?: EventMessageTypeNames;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageGeneric.ts:
```diff
@@ -0,0 +1,41 @@
+}
+
+export class EventMessageGeneric extends AbstractEventMessage {
+	readonly __type: string = EventMessageTypeNames.generic;
```
```suggestion
	readonly __type = EventMessageTypeNames.generic;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageAudit.ts:
```diff
@@ -0,0 +1,74 @@
+}
+
+export class EventMessageAudit extends AbstractEventMessage {
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+	readonly __type: string = EventMessageTypeNames.audit;
```
```suggestion
	readonly __type = EventMessageTypeNames.audit;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageNode.ts:
```diff
@@ -0,0 +1,49 @@
+}
+
+export class EventMessageNode extends AbstractEventMessage {
+	readonly __type: string = EventMessageTypeNames.node;
```
```suggestion
	readonly __type = EventMessageTypeNames.node;
```

netroy commented on packages/cli/src/eventbus/EventMessageClasses/EventMessageWorkflow.ts:
```diff
@@ -0,0 +1,61 @@
+}
+
+export class EventMessageWorkflow extends AbstractEventMessage {
+	readonly __type: string = EventMessageTypeNames.workflow;
```
```suggestion
	readonly __type = EventMessageTypeNames.workflow;
```

netroy commented on packages/workflow/src/MessageEventBus.ts:
```diff
@@ -0,0 +1,168 @@
+// ===============================
+
+export interface MessageEventBusDestinationOptions {
+	__type?: string;
```
```suggestion
	__type?: MessageEventBusDestinationTypeNames;
```

netroy commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestinationWebhook.ee.ts:
```diff
@@ -0,0 +1,372 @@
+	extends MessageEventBusDestination
+	implements MessageEventBusDestinationWebhookOptions
+{
+	__type: string = MessageEventBusDestinationTypeNames.webhook;
```
```suggestion
	__type = MessageEventBusDestinationTypeNames.webhook;
```

netroy commented on packages/cli/src/eventbus/MessageEventBusDestination/MessageEventBusDestination.ee.ts:
```diff
@@ -0,0 +1,125 @@
+	// static abstract deserialize(): MessageEventBusDestination | null;
+	readonly id: string;
+
+	__type: string;
```
```suggestion
	__type: MessageEventBusDestinationTypeNames;
```

netroy commented on packages/cli/package.json:
```diff
@@ -70,16 +70,21 @@
     "@types/convict": "^4.2.1",
     "@types/cookie-parser": "^1.4.2",
     "@types/express": "^4.17.6",
+    "@types/ioredis": "^4.0.0",
```
we upgrade ioredis last week. the updated package comes with it's own typings, and we should remove this stub package.

flipswitchingmonkey committed:
review changes

netroy committed:
Merge remote-tracking branch 'origin/master' into n8n-5359-mvp-event-bus

netroy committed:
delete duplicate lines from a bad merge

netroy committed:
decouple log-streaming UI options from public API

netroy committed:
Merge remote-tracking branch 'origin/master' into n8n-5359-mvp-event-bus

netroy committed:
logstreaming -> log-streaming for consistency

netroy committed:
do not make unnecessary api calls when log streaming is disabled

netroy committed:
prevent sentryClient.close() from being called if init failed

netroy committed:
fix the e2e test for log-streaming

flipswitchingmonkey committed:
review changes

flipswitchingmonkey committed:
cleanup

netroy committed:
use `private` for one last private property

netroy committed:
do not use node prefix package names.. just yet

netroy committed:
remove unused import

netroy committed:
fix the tests

because there is a folder called `events`, tsc-alias is messing up all imports for native events module.
https://github.com/justkey007/tsc-alias/issues/152
